File: src/main/java/no/nav/tjenestepensjon/simulering/consumer/TpConfigConsumerService.java
package no.nav.tjenestepensjon.simulering.consumer;

import static no.nav.tjenestepensjon.simulering.config.CacheConfig.TP_ORDNING_LEVERANDOR_CACHE;
import static no.nav.tjenestepensjon.simulering.config.WebClientConfig.webClient;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import no.nav.tjenestepensjon.simulering.domain.TPOrdning;

@Service
public class TpConfigConsumerService implements TpConfigConsumer {

    private String tpConfigUrl;
    private WebClient webClient = webClient();

    @Value("${TP_CONFIG_URL}")
    public void setTpConfigUrl(String tpConfigUrl) {
        this.tpConfigUrl = tpConfigUrl;
    }

    @Override
    @Cacheable(value = TP_ORDNING_LEVERANDOR_CACHE)
    public String findTpLeverandor(TPOrdning tpOrdning) {
        return webClient.get()
                .uri(tpConfigUrl + "/tpleverandoer/" + tpOrdning.getTpId())
                .retrieve()
                .bodyToMono(String.class)
                .block();
    }
}

File: src/main/kotlin/no/nav/tjenestepensjon/simulering/consumer/TpConfigConsumerService.kt
package no.nav.tjenestepensjon.simulering.consumer

import no.nav.tjenestepensjon.simulering.config.CacheConfig.Companion.TP_ORDNING_LEVERANDOR_CACHE
import no.nav.tjenestepensjon.simulering.model.domain.TPOrdning
import org.springframework.beans.factory.annotation.Value
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.bodyToMono

@Service
class TpConfigConsumerService(
        @Value("\${TP_CONFIG_URL}") val tpConfigUrl: String,
        val webClient: WebClient
) : TpConfigConsumer {

    @Cacheable(TP_ORDNING_LEVERANDOR_CACHE)
    override fun findTpLeverandor(tpOrdning: TPOrdning): String = webClient.get()
            .uri("$tpConfigUrl/tpleverandoer/${tpOrdning.tpId}")
            .retrieve()
            .bodyToMono<String>()
            .block()!!
}